src/System.Private.Uri/src/System/IPv6AddressHelper.cs

```diff
@@ -15,68 +14,86 @@ internal static class IPv6AddressHelper
         // fields
 
         private const int NumberOfLabels = 8;
-        // Lower case hex, no leading zeros
-        private const string CanonicalNumberFormat = "{0:x}";
-        private const string EmbeddedIPv4Format = ":{0:d}.{1:d}.{2:d}.{3:d}";
-        private const char Separator = ':';
 
         // methods
 
-        internal static string ParseCanonicalName(string str, int start, ref bool isLoopback, ref string scopeId)
+        internal static unsafe string ParseCanonicalName(string str, int start, ref bool isLoopback, ref string scopeId)
         {
-            unsafe
-            {
-                ushort* numbers = stackalloc ushort[NumberOfLabels];
-                // optimized zeroing of 8 shorts = 2 longs
-                ((long*)numbers)[0] = 0L;
-                ((long*)numbers)[1] = 0L;
-                isLoopback = Parse(str, numbers, start, ref scopeId);
-                return '[' + CreateCanonicalName(numbers) + ']';
-            }
-        }
+            ushort* numbers = stackalloc ushort[NumberOfLabels];
+            // optimized zeroing of 8 shorts = 2 longs
+            ((long*)numbers)[0] = 0L;
+            ((long*)numbers)[1] = 0L;
+            isLoopback = Parse(str, numbers, start, ref scopeId);
 
-        internal static unsafe string CreateCanonicalName(ushort* numbers)
-        {
             // RFC 5952 Sections 4 & 5 - Compressed, lower case, with possible embedded IPv4 addresses.
 
             // Start to finish, inclusive.  <-1, -1> for no compression
             KeyValuePair<int, int> range = FindCompressionRange(numbers);
             bool ipv4Embedded = ShouldHaveIpv4Embedded(numbers);
 
-            StringBuilder builder = new StringBuilder();
+            Span<char> stackSpace = stackalloc char[48]; // large enough for any IPv6 string, including brackets
```

Comment:
What about link-local addresses that have the %interface-name in the address?  Is there enough room for that?

Reply:
This method doesn't output the %interface-name.  Where is that done?

Reply:
>This method doesn't output the %interface-name. Where is that done?

I wasn't sure if it does or not.  But I was reacting to this line and comment where you said it would be enough for any IPv6 address:

```c#
stackalloc char[48]; // large enough for any IPv6 string, including brackets
```

So, I just wanted to make sure that link-local addresses with %interfacename was considered as part of this design change.  If link-local addresses aren't used here, then this is ok.

Reply:
I think link-local addresses are rare and it would be ok if processing is more expensive. 
Also note that on Unix the interface identifier can be string: https://github.com/dotnet/corefx/issues/27529

Reply:
If link-local addresses can't be present, then we don't need any 'if' check for a slower way to process this.  But if link-local addresses can be in this part of the code, we need to handle it, or we can overflow the stack buffer.

Reply:
This method does not format the %interface-name portion of link-local addresses.